pkg/health/condition/check_backup_ready.go

```diff
@@ -37,101 +38,194 @@ const (
-			result.reason = BackupSucceeded
-			result.message = "Delta snapshot backup succeeded"
-			result.status = druidv1alpha1.ConditionTrue
-			return result
-		}
-	} else if deltaLeaseRenewTime == nil && fullLeaseRenewTime != nil {
-		//Most probable during a startup scenario for new clusters
-		//Special case. Return Unknown condition for some time to allow delta backups to start up
-		if time.Since(fullLeaseRenewTime.Time) > 5*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-			result.message = "Periodic delta snapshots not started yet"
-			return result
-		}
-	} else if deltaLeaseRenewTime != nil && fullLeaseRenewTime != nil {
-		//Both snap leases are maintained. Both are expected to be renewed periodically
-		if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration && time.Since(fullLeaseRenewTime.Time) < 24*time.Hour {
-			result.reason = BackupSucceeded
-			result.message = "Snapshot backup succeeded"
-			result.status = druidv1alpha1.ConditionTrue
-			return result
-		}
-	}
-
-	//Cases where snapshot leases are not updated for a long time
-	//If snapshot leases are present and leases aren't updated, it is safe to assume that backup is not healthy
-
-	if etcd.Status.Conditions != nil {
-		var prevBackupReadyStatus druidv1alpha1.Condition
-		for _, prevBackupReadyStatus = range etcd.Status.Conditions {
-			if prevBackupReadyStatus.Type == druidv1alpha1.ConditionTypeBackupReady {
-				break
-			}
-		}
-
-		// Transition to "False" state only if present state is "Unknown" or "False"
-		if deltaLeaseRenewTime != nil && (prevBackupReadyStatus.Status == druidv1alpha1.ConditionUnknown || prevBackupReadyStatus.Status == druidv1alpha1.ConditionFalse) {
-			if time.Since(deltaLeaseRenewTime.Time) > 3*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-				result.status = druidv1alpha1.ConditionFalse
-				result.reason = BackupFailed
-				result.message = "Stale snapshot leases. Not renewed in a long time"
-				return result
-			}
-		}
-	}
-
-	//Transition to "Unknown" state is we cannot prove a "True" state
-	return result
 }
 
-func getDeltaSnapLeaseName(etcd *druidv1alpha1.Etcd) string {
-	return fmt.Sprintf("%s-delta-snap", string(etcd.ObjectMeta.Name))
+// handleSnapshotterStartupCase handles cases where snapshotter has just started,
+// so only full snapshot has been taken, while delta snapshot is still not taken.
+// Returns `Unknown` condition for some time to allow delta snapshotting to begin, because
+// even though the full snapshot may have succeeded within the required time, we must still wait
+// for delta snapshotting to begin to consider the backups as healthy, to maintain the given RPO.
+func handleSnapshotterStartupCase(fullSnapshotLeaseRenewTime time.Time, deltaSnapshotRenewalGracePeriod time.Duration) *result {
+	if time.Since(fullSnapshotLeaseRenewTime) > deltaSnapshotRenewalGracePeriod {
+		return createBackupConditionResult(
+			druidv1alpha1.ConditionFalse, BackupFailed,
```

Comment:
I am not entirely convinced if this results in `BackupFailed`. Can you semantically define this term to better gauge if this is the correct reason code? The reason is that you still have a full-snapshot that has been successfully backed-up. Should you call this `BackupFailed` or `DeltaSnapshotBackupFailed`?

Reply:
That again now depends on how granular we want our `Reason` codes to be. `BackupFailed` denotes that either of the backups failed, and either case is dangerous. Without a full snapshot taken on time, we risk longer restoration time and hence a longer RTO. And without a delta snapshot taken on time, we violate SLAs since RPOs is affected. So differentiating between which snapshot failed in the `Reason` does not provide much benefit to operators. If they want more info on why `BackupFailed` was set, they can look into the `Message`.